src/transformStateWithClones.js

```diff
@@ -7,7 +7,63 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCloneArray = [];
+  let indexLastActions = 0;
+
+  actions.forEach((action) => {
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case 'clear':
+        stateCloneArray.push(Object.assign({}));
+        break;
+
+      case 'addProperties':
+
+        if (!stateCloneArray.length) {
+          stateCloneArray.push(Object.assign({}, state, extraData));
+          break;
+        }
+
+        indexLastActions = stateCloneArray.length - 1;
+
+        stateCloneArray.push(
+          Object.assign({}, stateCloneArray[indexLastActions], extraData)
+        );
+
+        break;
+
+      default:
+        if (!stateCloneArray.length) {
```

Comment:
Again, no need to treat the first action different from others. You should always clone the last state and apply the action.